### ISSUE TITLE
Consider replacing `String` with `Text` or `ByteString` in most locations

### ISSUE THREAD

**Issue by rwe** [agda/agda#3741]
The Agda codebase leans strongly toward `String` v. `Text` or `ByteString`, although `String` has well-documented performance and memory-weight issues.

I have not profiled this, but given the amount of string operations and related types (e.g. `QName` → `Name` → `String`), it's possible that there may be some easy runtime performance gains by simply switching to `Text` or `ByteString` in most places.

References:
  * https://mmhaskell.com/blog/2017/5/15/untangling-haskells-strings
  * https://guide.aelve.com/haskell/strings-o62hqc69
  * https://wiki.haskell.org/Performance/Strings
  * https://www.reddit.com/r/haskell/comments/3rrlih/text_or_string/

**Comment by banacorn** [comment]
I think perhaps the *short* version of [`Text`](https://hackage.haskell.org/package/text-short-0.1.1/docs/Data-Text-Short.html) and [`ByteString`](https://hackage.haskell.org/package/bytestring-0.10.8.2/docs/Data-ByteString-Short.html) would even be better suited.

**Comment by jespercockx** [comment]
Seems like it would be worth trying out. Personally I would be a bit surprised if this has more than, say, 1% performance impact, but I've been surprised before :)

**Comment by nad** [comment]
I would prefer Text to ByteString.

**Comment by UlfNorell** [comment]
I would also be surprised if it gives any measurable performance gains, but it should be straightforward to try it out. We already have `OverloadedStrings` on by default which should make it require fewer changes.

**Comment by rwe** [comment]
I agree it's not likely to be earth-shattering (though, I wonder also if it might help with GC performance), just possibly low-hanging fruit worth noting. Easy for someone otherwise-unfamiliar with the internals to contribute.

**Comment by guillaumebrunerie** [comment]
Maybe it would improve memory usage when running Agda though? I’m guessing that a lot of the memory used by Agda consists of names of identifiers, but I may be wrong.

The second link in the OP contains the following comments:

> If you have lots of small strings, you can switch to text-short, which has less memory overhead than Text.
> If you have lots of small strings and you anticipate that many of them will be the same (e.g. if you're writing a compiler), you can use interned strings with something like intern, or roll your own like GHC did with its FastString.

**Comment by guillaumebrunerie** [comment]
Using the numbers from https://wiki.haskell.org/GHC/Memory_Footprint, for an ASCII string of length N on a 64 bit computer:
- a String uses 8 + 40N bytes
- a short Text uses 48 + N bytes

It seems to me there is a potentially huge gain here.

**Comment by nad** [comment]
There has already been some work on improving sharing of strings (by @andreasabel and Makoto Takeyama?), so I doubt that further work will make much of a difference.

**Comment by andreasabel** [comment]
In 2014 in the process of making serialization more efficient, I tried to get rid of `String`.  Remnants of this refactoring attempt are several aliases to String, e.g., this one:
https://github.com/agda/agda/blob/126cab169a4d9e80b46fba102e5c7252720459ca/src/full/Agda/Syntax/Internal.hs#L252-L262
See 5df4e94e46727923d4814492d00ad7bd139de542 .
My experiments did not demonstrate a clear improvement (unlike the other changes to serialization). Even if `String` takes more memory than `Text`, the strings coming from the Agda files are shared, and substitution (while destroying sharing of terms), does not destroy sharing of these strings.  Thus, I'd be surprised as well if much was to be gained here.  A further argument would be that the strings are all relatively short, as long atomic identifiers are the exception.  Qualified identifiers are anyway not stored as monolithic strings.
@guillaumebrunerie wrote:
> It seems to me there is a potentially huge gain here.

Do you hold up your opinion in sight of my arguments, @guillaumebrunerie ?

**Comment by guillaumebrunerie** [comment]
> Do you hold up your opinion in sight of my arguments, @guillaumebrunerie ?

Ok, you’re certainly more aware of the internals of Agda than I am.

Where does the sharing happen? Is it visible somewhere explicitly in the code?

I’m just wondering why they say that for a compiler you should use some special version of strings. If GHC has its own implementation of fast strings, why wouldn’t Agda benefit much from it?

**Comment by andreasabel** [comment]
No, sharing is automatic in Haskell, unless something destroys the sharing.  For instance, a concrete name contains a string / strings, but this concrete name is passed around and stored in other structures without modification of the string.  Thus, the string remains shared among all these references.
Also, identifier resolution is not using strings.  After scope checking, all names are unique integers (Abstract.Name).  Only the scope checker has to resolve identifiers by their concrete name (string).  But this pass is (ideally) O(n log n) in the file size n, as each identifier from the source code has to be resolved only once.  (Disregarding `record` hacks, see #434 and friends.)
